dalm/datasets/reading_comprehension_generation/synthetic_based.py

```diff
@@ -0,0 +1,104 @@
+    return [
+        {
+            "role": "system",
+            "content": "You are a helpful and meticulous instruction following question and answer making chatbot. Please refrain from acknowledgments, additions or niceties of any sort",
+        },
+        {"role": "user", "content": prompt},
+    ]
+
+
+def generate_synthetic_data(model_pipeline, text, generation_params):
+    prompt = gen_prompt(text)
+    prompt = model_pipeline.tokenizer.apply_chat_template(prompt, tokenize=False, add_generation_prompt=True)
+    outputs = model_pipeline(prompt, **generation_params)
+
+    return outputs[0]["generated_text"]
+
+
+def generate_synthetic_dataset(
+    model_name,
+    input_directory,
+    state_file=None,
+    chunk=False,
+    context_length=2048,
+    generation_params={
+        "max_new_tokens": 600,
+        "do_sample": True,
+        "temperature": 0.7,
+        "top_k": 5,
+        "top_p": 0.95,
+        "return_full_text": False,
+    },
+):
+    model_pipeline = pipeline("text-generation", model=model_name, torch_dtype=torch.bfloat16, device_map="auto")
+
+    input_files = list_dir(input_directory)
+
+    if chunk:
+        tokenizer = model_pipeline.tokenizer
+        tokens = tokenizer.apply_chat_template(gen_prompt(""), tokenize=False, add_generation_prompt=True)
+        CONSTANT = len(tokenizer(tokens)["input_ids"])
+        k = context_length - CONSTANT
+
+    if state_file:
+        if os.path.exists(state_file):
+            with open(state_file, "rb") as f:
+                state = pickle.load(f)
+        else:
+            state = {"processed_files": []}
+            pickle.dump(state, open(state_file, "wb"))
+
+    for file, text in input_files:
+        if state and file in state["processed_files"]:
+            continue
+
+        if chunk:
+            for chunk_ in text_chunker(text, tokenizer, k):
+                gen_text = generate_synthetic_data(model_pipeline, chunk_, generation_params)
+                yield gen_text, chunk_
+        else:
+            gen_text = generate_synthetic_data(model_pipeline, text, generation_params)
+            yield gen_text, text
+
+        if state:
+            state["processed_files"].append(file)
+            pickle.dump(state, open(state_file, "wb"))
+
+
+if __name__ == "__main__":
+    parser = argparse.ArgumentParser()
+    parser.add_argument("--model_name", type=str, required=True)
+    parser.add_argument("--input_directory", type=str, required=True)
+    parser.add_argument("--output_directory", type=str, required=True)
+    parser.add_argument("--state_file", type=str, required=False, default="rc_generation_state.pkl")
+    parser.add_argument("--context_length", type=int, default=2048)
+    parser.add_argument("--chunk", action="store_true")
+
+    args = parser.parse_args()
+
+    """
+    Pipeline here includes chunking, generation and parsing of question and answer into a list of exchanges
+    that can be used directly for training
```

Comment:
I tried generating a dataset and using it for training, but I'm getting errors.  The generated dataset has the following files:

```
root@04e369b03ae7:/# ls -alh train_dataset/data/
total 40K
drwxr-xr-x 2 root root  115 Nov 20 14:51 .
drwxr-xr-x 3 root root   26 Nov 20 14:52 ..
-rw-r--r-- 1 root root 7.7K Nov 20 14:51 gen_0.txt
-rw-r--r-- 1 root root 8.9K Nov 20 14:51 gen_1.txt
-rw-r--r-- 1 root root 6.6K Nov 20 14:51 gen_2.txt
-rw-r--r-- 1 root root 6.3K Nov 20 14:51 gen_3.txt
-rw-r--r-- 1 root root 3.0K Nov 20 14:51 gen_5.txt
```

but when I try to train on it with `python -m dalm.training.generator_only.trainer --model_name "meta-llama/Llama-2-7b-chat-hf" --dataset_name train_dataset --size_valid_set 1`, it gives error:

```
FileNotFoundError: Directory train_dataset is neither a `Dataset` directory nor a `DatasetDict` directory.
```

What changes need to be made to the dataset for it to be loaded?

Reply:
Also it didn't work if I tried to use the default dataset:

```
python -m dalm.training.generator_only.trainer --model_name "meta-llama/Llama-2-7b-chat-hf"  --size_valid_set 100
```

It gives error: `FileNotFoundError: Directory arcee-ai/azure-reading-comprehension-dataset not found`

Looks like the code is defaulting `local_dataset: bool = True` which doesn't work with the default argument for the the dataset name: `arcee-ai/azure-reading-comprehension-dataset`

Reply:
All good now, I had an invalid HF dataset in the directory and it didn't recognize it.  I since switched to the `reading_comprehension_pipeline` for testing which is working now.